### ISSUE TITLE
All tested fonts the 'space' character has a very large Y cropping value.

### ISSUE THREAD

**Issue by willmotil** [MonoGame/MonoGame#6545]
Ok so because i was doing something a little unusual basically related to word wrapping and editing.
The function i was making had to detect were a click was in a word wrapped message. For editing purposes it had to translate that to a character index in the original. 
Were in i discover a strange discrepancy in the first couple tests i ran. 

So i hunted it down.

Every font i tested had its white space characters cropping.Y.
Value at a strangely large amount in most case by about two lines down.

![a odd space bug](https://user-images.githubusercontent.com/7495196/48636097-ef5d4200-e997-11e8-968b-9414fcc179be.png)

I temporarily fixed it in my function by seting any spaces y cropping to zero but thats not a real solution.

I believe this is related to this but im not sure how.

https://github.com/MonoGame/MonoGame/blob/6f34eb393aa0ac005888d74c5c4c6ab5615fdc8c/MonoGame.Framework.Content.Pipeline/Graphics/Font/GlyphCropper.cs

```
			// Crop the top.
			while ((glyph.Subrect.Height > 1) && BitmapUtils.IsAlphaEntirely(0, glyph.Bitmap, new Rectangle(glyph.Subrect.X, glyph.Subrect.Y, glyph.Subrect.Width, 1)))
			{
				glyph.Subrect.Y++;
				glyph.Subrect.Height--;

				glyph.YOffset++;
}
```

http://community.monogame.net/t/font-spacing-issue/11030/4

MonoGame 3.7
Windows
DesktopGL, WindowsDX

**Comment by mrhelmut** [comment]
I believe it's somehow related to this code, and partly related to how freetype generates glyph rectangles.

I'm assuming that you are loading a font through a SpriteFont descriptor, and not a bitmap font?

I think that freetype already gets a fully cropped rectangle when reading the space glyph info, and then MonoGame tries to do the same thing which may be useful for bitmap fonts, but is twice the same job for non-bitmap fonts (hence the YOffset being too far off).
(But if you're working with bitmap fonts, I'm probably wrong.)

Besides this discrepancy, a totally flat rectangle for glyphs with no pixel is what a typography system is expecting, the YOffset shouldn't be changed compared to what freetype is reading (this basically comes from the font info itself).
I guess that what you would have liked is that a space glyph would still have a positive height (instead of 0) so you could have a collider. But that would probably break some font renderers, the YOffset value is an essential data in font rendering accuracy (even in the case of a totally flat glyph, some circumstances may require it).

It's true that we're talking about spaces. It's a character that shouldn't have rendering issues with a wrong YOffset (and doesn't seem to have any issue with this double cropping bug). So maybe we could hack something around? Yet, I don't know if we should hack around fonts for a specific use, it may still break things we're not aware of.

Maybe you could rebuild the SpriteFont by hacking the space glyph to allow your collision detection to work?

Anyway, I'm going to take a look into this and at least fix the far too off YOffset issue (it shouldn't be more than flat).

**Comment by willmotil** [comment]
> I'm assuming that you are loading a font through a SpriteFont descriptor, and not a bitmap font?

Yes i typically just add a font to my system then changed the descriptor to which ever font i want to use or test with. Makes things fast and simple. However i have a bitmap font i made myself and added to my system, when i import that as a spritefont it also gets that big y cropping value. So the importing is doing something funky there.

> the YOffset shouldn't be changed compared to what freetype is reading (this basically comes from the font info itself). I guess that what you would have liked is that a space glyph would still have a positive height (instead of 0) so you could have a collider.

In my first couple tests i just naively checked against the drawing rectangle position and realized something weird was going on i zeroed out the y cropping when i realized i couldn't click any spaces.

Fortunately i can just checked the click point against the x y offsets as that is actually tracking the real start positions separate from any cropping kerning ect so the collision isn't a problem anyways.

But 
I went back and checked what exactly was going on in the first place as i could see it was like really off as clicks were bouncing around on white space by huge amounts. I remember the conversation i was having with someone else on the forum. 

This person was however having some problems with his white spacing being different to xna shown in a image in that link. 

http://community.monogame.net/uploads/default/original/2X/0/044f19286f6ae3314da2142b63483b03eec5314b.png

http://community.monogame.net/t/font-spacing-issue/11030

I was thinking this maybe the cause. If it is pulling in bogus y cropping values the other values could be getting messed up too.

**Comment by mrhelmut** [comment]
Ah yes, good catch, there may be an issue with cropping the XOffset or XAdvance values as well!
